fix: Updates to SSL & Custom domain doc (#1002)
appsmithorg/appsmith-docs: website/docs/getting-started/setup/instance-configuration/custom-domain/README.md

---
sidebar_position: 7
---

> # SSL for a Custom Domain

Reviewer: ```suggestion
# Custom Domain and SSL
```

> You can configure SSL for your Custom Domain to establish secured connections. This page provides instructions to set up SSL for your Custom Domain on your Appsmith instance.

Reviewer: ```suggestion
This page provides instructions to configure SSL for your custom domain to establish secured connections on your Appsmith instance.
```

Author: This sentence becomes too difficult to read. Will reword to:
`This page explains how to set up SSL for your custom domain on the Appsmith instance. So that you can make secure connections.`

> <VideoEmbed host="youtube" videoId="0llo1exi4IY" title="How To Self-Host Appsmith With A Custom Domain" caption="Set up SSL and HTTPS for your Custom Domain on your self-hosted Appsmith Instance"/>

Reviewer: ```suggestion
<VideoEmbed host="youtube" videoId="0llo1exi4IY" title="How To Self-Host Appsmith With A Custom Domain" caption="Set up SSL for your Custom Domain on your self-hosted Appsmith Instance"/>
```

## Prerequisites
Before configuring SSL for your custom domain, make sure you have the following:

> 1. A self-hosted Appsmith instance running. For more information about installing Appsmith, see [Installation Guides](/getting-started/setup/installation-guides)

Reviewer: ```suggestion

```

2. A domain name - You can get a custom domain from popular providers like [GoDaddy](https://in.godaddy.com/help/create-a-subdomain-4080), [Amazon Route 53](https://aws.amazon.com/premiumsupport/knowledge-center/create-subdomain-route-53/), [Digital Ocean](https://www.digitalocean.com/docs/networking/dns/how-to/add-subdomain/), [NameCheap](https://www.namecheap.com/support/knowledgebase/article.aspx/9776/2237/how-to-create-a-subdomain-for-my-domain), and [Domain.com](https://www.domain.com/help/article/domain-management-how-to-update-subdomains).
3. Ports 80 and 443 are open and accessible.

> ## Configure SSL

Reviewer: ```suggestion
## Configure SSL with Let's Encrypt
```

> Once you have a Custom Domain, you can choose to set up SSL in one of the following ways:

Reviewer: ```suggestion
```

> * [Generate and Configure SSL](#generate-and-configure-ssl)

Reviewer: remove both these points

* [Configure a Custom SSL Certificate](#configure-a-custom-ssl)

> ## Generate and configure SSL

Reviewer: remove this heading

> On Appsmith, you can choose to add your Custom Domain by using one of the following ways:

Reviewer: ```suggestion
On Appsmith, you can add your custom domain by using one of the following ways:
```

* [Admin settings](#admin-settings)
* [Environment Variables](#environment-variables)

### Admin settings

> You can use Admin Settings User Interface to set up a custom domain. The **Custom Domain** property is available under **Advanced** Settings. Add the domain name to the Custom Domain setting and click the **SAVE & RESTART** button. The Custom domain setup can be done using the Custom Domain setting for all your installation types except Kubernetes. For more information about how to set up TLS on Kubernetes, see [How to configure TLS for Appsmith Kubernetes installation](/getting-started/setup/installation-guides/kubernetes#configure-tls).

Reviewer: ```suggestion
You can use [Admin Settings](link) page to set up a custom domain. The **Custom Domain** property is available under **Advanced** Settings. Add the domain name to the Custom Domain setting and click the **SAVE & RESTART** button. 

Should the below be a callout?
The Custom domain setup can be done using the Custom Domain setting for all your installation types except Kubernetes. For more information about how to set up TLS on Kubernetes, see [How to configure TLS for Appsmith Kubernetes installation](/getting-started/setup/installation-guides/kubernetes#configure-tls).
```

Author: Added a callout

Author: I've added a callout at the start of the page so that users don't have to read until the Admin settings section and realize they need to navigate to another page for Kubernetes. Also, I kept the para as is in case someone misses the callouts and comes here.

<figure>
  <img src="/img/setup-custom-domain-using-admin-settings.png" style= {{width:"700px", height:"auto"}} alt="Setup your custom domain using Admin settings"/>
  <figcaption align = "center"><i>Setup your custom domain using Admin settings</i></figcaption>
</figure>

### Environment variables

> Appsmith is deployed on a Docker container. To configure your custom domain, you can use the environment variable `APPSMITH_CUSTOM_DOMAIN` in a `docker.env` file. Follow the steps below:

Author: As per my discussion with @sum35h - We wanted to provide this information to users that Appsmith uses Docker, and so platforms that support docker will also have similar configuration. The difference could be in the location of variables. CC: @sum35h

1. Navigate to the `docker.env` file located in the installation root folder. For example, if you are using Docker installation, you can locate the file in the `<PROJECT_ROOT>/stacks/configuration` file. Similarly, the file is in the `/home/ubuntu/appsmith/stacks/configuration` folder on AWS AMI.
    1. Update the key `APPSMITH_CUSTOM_DOMAIN` as shown below:

        ```yaml
        APPSMITH_CUSTOM_DOMAIN=<ADD_CUSTOM_DOMAIN_HERE>
        ```
    2. Restart the Appsmith container by using the following command:

        ```bash
        docker-compose restart appsmith
        ```

> When you restart Appsmith, it generates an SSL certificate for your custom domain. Appsmith uses [Let's Encrypt](https://letsencrypt.org) to generate and maintain the certificate so that it remains up to date. Once the restart is successful, you can use your custom domain mapped to port 443 via HTTPS to access Appsmith in your browser.

Reviewer: this para should be in the beginning

> ## Configure a Custom SSL

Reviewer: ```suggestion
## Configure custom SSL
```

If you already have SSL Certificate and want to use that, follow the steps listed below:

* Rename the certificate file as `fullchain.pem` and key file as `privkey.pem`.
* Copy these files into the subdirectory `<MOUNTING-DIRECTORY>/ssl/`. Ensure that you change `<MOUNTING-DIRECTORY>` by the mounting volume directory available in the `docker-compose.yml`. For example, the default value is `./stacks`.
* Add your Custom Domain to the environment variable `APPSMITH_CUSTOM_DOMAIN` in the 'docker.env' file.
* Restart the container using `docker restart appsmith`

Provided that the certificate files are present in the `<MOUNTING-DIRECTORY>/ssl` folder, the container uses them.

:::note
If you want to configure a Custom SSL on your Heroku or DigitalOcean Appsmith installation platform, follow the steps as listed in the guides below:
* [How to Configure SSL for your Heroku dyno](https://devcenter.heroku.com/articles/ssl)
* [How do I install an SSL Certificate on a DigitalOcean Droplet](https://docs.digitalocean.com/support/how-do-i-install-an-ssl-certificate-on-a-droplet/)
:::

## Troubleshooting

> If you’re having issues accessing Appsmith after Appsmith SSL Configuration, please see the [Unable to Access Appsmith](/help-and-support/troubleshooting-guide/deployment-errors#unable-to-access-appsmith) troubleshooting guide. If you continue to have problems reach out on [Discord Server](https://discord.com/invite/rBTTVJp) or [send an email to support](mailto:support@appsmith.com) or ask questions on the [community forum](https://community.appsmith.com/).

Reviewer: make sure troubleshooting link is specific to custom domain or SSL issues

Author: I confirmed with @sum35h that this link works for a custom domain and SSL issue fix.

> ## Further reading

Reviewer: system logs and instance management

    1. Update the key `APPSMITH_CUSTOM_DOMAIN` as shown below:

        ```yaml
        APPSMITH_CUSTOM_DOMAIN=<ADD_CUSTOM_DOMAIN_HERE>
        ```
    2. Restart the Appsmith container by using the following command:

        ```bash
        docker-compose restart appsmith
        ```

On restart Appsmith generates SSL certificate for your custom domain. Appsmith uses [Let's Encrypt](https://letsencrypt.org) to generate and keep the certificate up to date. After a successful restart, you can use the custom domain that's mapped to port 443 using HTTPS to access Appsmith in the browser.

### Configure a Custom SSL
If you already have SSL Certificate and want to use that, follow the steps listed below:

* Rename the certificate file as `fullchain.pem` and key file as `privkey.pem`.
* Copy these files into the subdirectory `<MOUNTING-DIRECTORY>/ssl/`. Ensure that you change `<MOUNTING-DIRECTORY>` by the mounting volume directory available in the `docker-compose.yml`. For example, the default value is `./stacks`.

> * Restart the container using `docker restart appsmith`

Author: Done.
